xds/internal/xdsclient/channel_test.go

```diff
@@ -0,0 +1,787 @@
+	nodeID := uuid.New().String()
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultClientListener(listenerResourceName, routeConfigurationName)},
+		SkipValidation: true,
+	}
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Create an xdsChannel for the test with a long watch expiry timer
+	// to ensure that watches don't expire for the duration of the test.
+	xc := xdsChannelForTest(t, mgmtServer.Address, nodeID, 2*defaultTestTimeout)
+	defer xc.close()
+
+	// Subscribe to the resource created above and verify that a request is sent
+	// for the same.
+	xc.subscribe(listenerType, listenerResourceName)
+	if err := waitForResourceNames(ctx, t, ldsResourcesCh, []string{listenerResourceName}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Wait for the above resource to be ACKed.
+	if err := waitForResourceNames(ctx, t, ldsResourcesCh, []string{listenerResourceName}); err != nil {
+		t.Fatal(err)
+	}
+
+	// Unsubscribe to the resource created above.
+	xc.unsubscribe(listenerType, listenerResourceName)
+	if err := waitForResourceNames(ctx, t, ldsResourcesCh, nil); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// Tests the load reporting functionality of the xdsChannel.  It creates an
+// xdsChannel, starts load reporting, and verifies that an LRS streaming RPC is
+// created. It then makes another call to the load reporting API and ensures
+// that a new LRS stream is not created. Finally, it cancels the load reporting
+// calls and ensures that the stream is closed when the last call is canceled.
+//
+// Note that this test does not actually report any load. That is already tested
+// by an e2e style test in the xdsclient package.
+func (s) TestChannel_LRS_ReportLoad(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	// Create a management server that serves LRS.
+	mgmtServer := e2e.StartManagementServer(t, e2e.ManagementServerOptions{SupportLoadReportingService: true})
+
+	// Create an xdsChannel for the test. Node id and watch expiry timer don't
+	// matter for LRS.
+	xc := xdsChannelForTest(t, mgmtServer.Address, "", defaultTestTimeout)
+	defer xc.close()
+
+	// Start load reporting and verify that an LRS streaming RPC is created.
+	_, stopLRS1 := xc.reportLoad()
+	lrsServer := mgmtServer.LRSServer
+	if _, err := lrsServer.LRSStreamOpenChan.Receive(ctx); err != nil {
+		t.Fatalf("Timeout when waiting for an LRS streaming RPC to be created: %v", err)
+	}
+
+	// Make another call to the load reporting API, and ensure that a new LRS
+	// stream is not created.
+	_, stopLRS2 := xc.reportLoad()
+	sCtx, sCancel := context.WithTimeout(context.Background(), defaultTestShortTimeout)
+	defer sCancel()
+	if _, err := lrsServer.LRSStreamOpenChan.Receive(sCtx); err != context.DeadlineExceeded {
```

Comment:
Isn't this testing the LRS Streaming reference counting functionality, which is a layer below this one? All this layer does is check if closed and forward down. Isn't it more appropriate for this functionality to be tested at the LRS Stream layer where ref counting is implemented with respect to creating on the first ref count etc.

Reply:
Ok I still feel like it's a big test/maintenance burden for what it's actually testing, but I buy both your points.